src/infra/src/table/folders.rs

```diff
@@ -0,0 +1,107 @@
+use config::meta::folder::Folder;
+use sea_orm::{
+    ActiveModelTrait, ActiveValue::NotSet, ColumnTrait, DatabaseConnection, EntityTrait,
+    ModelTrait, QueryFilter, QueryOrder, Set,
+};
+
+use super::entity::folder::{ActiveModel, Column, Entity, Model};
+use crate::{
+    db::{connect_to_orm, ORM_CLIENT},
+    errors,
+};
+
+// Flags that indicate the type of data the folder can contain.
+const DASHBOARDS_FOLDER_TYPE: i16 = 0;
+
+impl From<Model> for Folder {
+    fn from(value: Model) -> Self {
+        Self {
+            folder_id: value.folder_id,
+            name: value.name,
+            description: value.description.unwrap_or_default(),
+        }
+    }
+}
+
+/// Gets a folder by its `folder_id` surrogate key.
+pub async fn get(org_id: &str, folder_id: &str) -> Result<Option<Folder>, errors::Error> {
+    let client = ORM_CLIENT.get_or_init(connect_to_orm).await;
+    let folder = get_model(client, org_id, folder_id)
+        .await
+        .map(|f| f.map(Folder::from))?;
+    Ok(folder)
+}
+
+/// Lists all dashboard folders.
+pub async fn list_dashboard_folders(org_id: &str) -> Result<Vec<Folder>, errors::Error> {
+    let client = ORM_CLIENT.get_or_init(connect_to_orm).await;
+    let folders = list_models(client, org_id, DASHBOARDS_FOLDER_TYPE)
+        .await?
+        .into_iter()
+        .map(Folder::from)
+        .collect();
+    Ok(folders)
+}
+
+/// Creates a new folder or updates an existing folder in the database. Returns
+/// the new or updated folder.
+pub async fn put(org_id: &str, folder: Folder) -> Result<Folder, errors::Error> {
+    let client = ORM_CLIENT.get_or_init(connect_to_orm).await;
+    let folder_id = folder.folder_id;
+    let mut active: ActiveModel = match get_model(client, org_id, &folder_id).await? {
+        Some(model) => model.into(),
+        None => ActiveModel {
+            id: NotSet,
+            // We should probably set folder_id here rather than in caller code.
+            ..Default::default()
```

Comment:
Currently the `folder_id` foreign key is generated in the Service logic [here](https://github.com/openobserve/openobserve/blob/1cc4097c7aa155fb6b37f3ce3cbddb6bbaf451b7/src/service/dashboards/folders.rs#L58).

Personally, I think it would be better to generate `folder_id` here, so that that the Service logic doesn't have to worry about it - similarly to how auto-incremented primary keys are generated by the database.

However I did not move the key generation here in this PR because this PR is already very large and I would like to keep it focused solely on the changes necessary to migrate to the new `folder` table. If this is a change we'd like to do in the future then I'm happy to make it.